### ISSUE TITLE
Unexpected early stopping behaviour

### ISSUE THREAD

**Issue by ecly** [modernmt/modernmt#509]
I'm currently working on finding a good configuration for training with early stopping using ModernMT (which you may have seen from #508).

In doing so, I am getting somewhat confused with the current loss values reported in the logs, and the loss values used for early stopping. In one instance I train a network with the following command:
```bash
./mmt create -e DEEN_early_stop_fp16 de en data/DEEN/ \
  --arch transformer_mmt_base \
  --max-tokens 4096 \
  --seed 42 \
  --fp16 \
  --checkpoints-num 2 \
  --save-interval 1 \
  --validate-interval 1 \
  --no-epoch-checkpoints \
  --save-interval-updates 10000
```
And now here is my unexpected behaviour in the form of an excerpt from the 'training.log'.
```
...
| epoch 011 | valid on 'valid' subset | loss 3.337 | nll_loss 1.518 | ppl 2.86 | num_updates 90000 | best_loss 3.33733
| epoch 012 | valid on 'valid' subset | loss 3.321 | nll_loss 1.504 | ppl 2.84 | num_updates 100000 | best_loss 3.3213
| epoch 013 | valid on 'valid' subset | loss 3.305 | nll_loss 1.492 | ppl 2.81 | num_updates 110000 | best_loss 3.30541
| epoch 014 | valid on 'valid' subset | loss 3.298 | nll_loss 1.481 | ppl 2.79 | num_updates 120000 | best_loss 3.29849
| epoch 016 | valid on 'valid' subset | loss 3.290 | nll_loss 1.477 | ppl 2.78 | num_updates 130000 | best_loss 3.2899
2019-10-01 23:55:08,016 [INFO] - Stop criterion: current_loss = 3.671257, previous_loss = 3.655064
2019-10-01 23:55:08,017 [INFO] - Training interrupted by termination policy: validation loss has reached its plateau
```

The [criterion](https://github.com/modernmt/modernmt/blob/master/cli/train.py#L76) is using a loss window for previous/current loss, based on the `--checkpoints-num` arg, which in my case is 2. From this implementation, I still expect to see much lower values used for the criterion, and I similarly would not expect it to stop training, as I see the validation loss steadily decreasing.

Is my combination of options producing some unintended behaviour, or is validation loss from the tensorboard events somehow not aligned with the validation loss reported in the training.log?

From other experiments without early-stopping I would expect this setup to train for almost twice as long before reaching a validation plateau.

**Comment by ecly** [comment]
By opening tensorboard and looking at the reported validation loss values, the one that in tensorboard aligns with what fairseq reports as 'loss' is confusingly called 'best_loss' in tensorboard.

Inspecting fairseq's code, it looks like 'best_loss' is in fact the way to go, from [this snippet](https://github.com/pytorch/fairseq/blob/master/train.py#L255), where we see that 'best_loss' is what actually corresponds to the average validation loss. I'm gonna run an experiment with this now, and will return with a PR if this turns out to be the solution, and that you agree that it is.

**Comment by ecly** [comment]
[Here](https://github.com/pytorch/fairseq/issues/737) is a ticket from fairseq basically explaining the difference between loss and best_loss. They say they are refactoring, but for the time being, 'best_loss' is the running average for the epoch, whereas 'loss' is just some point, with the reasoning of it being smoothed within tensorboard. This means that we really want to use 'best_loss' as a representation for our validation loss. 

I will continue my tests for this to verify that my changes are working and submit a PR within a few days.

**Comment by ecly** [comment]
This issue was resolved in #511
